src/main/java/org/logstash/beats/ConnectionHandler.java

```diff
@@ -0,0 +1,112 @@
+package org.logstash.beats;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import io.netty.util.AttributeKey;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Manages the connection state to the beats client.
+ */
+public class ConnectionHandler extends ChannelDuplexHandler {
+    private final static Logger logger = LogManager.getLogger(ConnectionHandler.class);
+
+    public static AttributeKey<AtomicBoolean> CHANNEL_SEND_KEEP_ALIVE = AttributeKey.valueOf("channel-send-keep-alive");
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+        ctx.channel().attr(CHANNEL_SEND_KEEP_ALIVE).set(new AtomicBoolean(false));
+        if (logger.isTraceEnabled()) {
+            logger.trace("{}: channel activated", ctx.channel().id().asShortText());
+        }
+        super.channelActive(ctx);
+    }
+
+    /**
+     * {@inheritDoc}
+     * Sets the flag that the keep alive should be sent. {@link BeatsHandler} will un-set it. It is important that this handler comes before the {@link BeatsHandler} in the channel pipeline.
+     * Note - For large payloads, this method may be called many times more often then the BeatsHandler#channelRead due to decoder aggregating the payload.
+     */
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ctx.channel().attr(CHANNEL_SEND_KEEP_ALIVE).get().set(true);
+        if (logger.isDebugEnabled()) {
+            logger.debug("{}: batches pending: {}", ctx.channel().id().asShortText(),ctx.channel().attr(CHANNEL_SEND_KEEP_ALIVE).get().get());
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br/>
+     * <p>
+     * <strong>IdleState.WRITER_IDLE <br/></strong>
```

Comment:
I know this is existing code already, but any idea on why using the WRITER_IDLE vs the READER_IDLE?

Reply:
A WRITER_IDLE event is emitted by the `IdleStateHandler` once the idle timeout is breached when writing back to the channel. In this case, the last time that an acknowledgement was sent to the Beats client - that acknowledgement could be the last sequence number, or a keep alive. By using the WRITER_IDLE event, we can make sure that a `keep alive` message is sent every 5 seconds

Reply:
makes sense, thanks